### PR TITLE
Localization upgrade

### PR DESCRIPTION
**Pull request by ArishSultan**
Implement new Localization pattern as described in version 1.22

### CODE REVIEW

**Comment by ArishSultan** [comment]
How to fix the cla issue?

**Comment by iapicca** [comment]
@ArishSultan 
https://github.com/canonical/ubuntu-desktop-installer/issues/8#issuecomment-773510447

**Comment by ArishSultan** [comment]
I have signed the CLA, should I add a new committ?

**Comment by MarcusTomlinson** [comment]
> I have signed the CLA, should I add a new committ?

Hi @ArishSultan, unfortunately our CLA check only works for those who've signed the CLA via Launchpad, and not GitHub as yet.

I've checked with Legal and they've confirmed you have signed the CLA, you (and whoever eventually reviews this) may ignore the cla-check status.

**Comment by oSoMoN** [comment]
Some functional regressions I noted when testing your branch:

 - On the welcome screen, the current locale is not selected by default (I use French on my desktop, if the French translations are available they should be loaded and the corresponding entry in the language list should be selected)
 
 - The "You may wish to read the release notes." string doesn't appear translated after manually selecting a language for which it is translated
 
- On the keyboard layout selection screen, the language for the current locale should be selected by default, if available

**Comment by ArishSultan** [comment]
Ok, i will fix these

**Comment by MarcusTomlinson** [comment]
> > I have signed the CLA, should I add a new committ?
> 
> Hi @ArishSultan, unfortunately our CLA check only works for those who've signed the CLA via Launchpad, and not GitHub as yet.
> 
> I've checked with Legal and they've confirmed you have signed the CLA, you (and whoever eventually reviews this) may ignore the cla-check status.

**Comment by Mejans** [comment]
Hello,

why would the Occitan be ignored o_Ô ?

**Comment by ArishSultan** [comment]
Its becuase the global material localization has not implementation for occitan,

You can see from the doxumwntation that the material package has implementation for 70 languages occitan is not included in those languages,

So, i ignored it for some time until someone provides its complete transalation.

**Comment by oSoMoN** [comment]
> Its becuase the global material localization has no implementation for occitan,

This shouldn't matter, we're not using Material translations in the app, are we? We shouldn't be tied to Flutter's list of translations. Ubuntu as a whole has translations for [244 languages](https://translations.launchpad.net/ubuntu/+translations) (not all of them are complete or of high quality though).

Can we make sure that all the existing translations are usable?

**Comment by Mejans** [comment]
From what I saw, Material translations is only about dates, months, and days things...
Doesn't seem to be used here yeah.

**Comment by ArishSultan** [comment]
Material Translation is used in whole application, most of the widgets inside the `material.dart` depend upon those translations i.e. AppBar.

For now I have implemented a dummy translation which will use `en` for all unknown translations.

**Comment by ArishSultan** [comment]
Why is this `subiquity_client` test failing?

**Comment by Mejans** [comment]
Thanks @ArishSultan it's appreciated

**Comment by ArishSultan** [comment]
anybody here?

**Comment by MarcusTomlinson** [comment]
> anybody here?

Sure. Don’t know why it’s failing. Hmm. I just ran CI against main again and passed fine. I’ll kick your build again and see if that resolves it. Otherwise I’ll have a look in Monday.

**Comment by MarcusTomlinson** [comment]
Subiquity tests are green again - interesting. Looks like there’s something up with the installer tests now though. Looks like a legit issue at least.

**Comment by ArishSultan** [comment]
Idk, test is passing on my side.

File: ubuntu_desktop_installer/lib/welcomepage.dart [hunk 1]
```diff
@@ -35,10 +35,10 @@ class _WelcomePageState extends State<WelcomePage> {
     _languageListScrollController = AutoScrollController();
 
     widget.client
-        .fetchKeyboardLayouts(kbdAssetName, Locale(Intl.defaultLocale));
+        .fetchKeyboardLayouts(kbdAssetName, UbuntuDesktopInstallerApp.locale);
     final locale = Intl.defaultLocale;
     for (var i = 0; i < widget.client.languagelist.length; ++i) {
-      if (widget.client.languagelist[i].item1.languageCode == locale) {
+      if (locale.contains(widget.client.languagelist[i].item1.languageCode)) {
```
Comment (MarcusTomlinson) [inline ubuntu_desktop_installer/lib/welcomepage.dart:41]: This is where the test is failing (fails locally for me as well). It appears that `locale` is null at this point.

**Review by MarcusTomlinson — CHANGES_REQUESTED** [review]

File: ubuntu_desktop_installer/lib/welcomepage.dart (hunk 1 above)
Comment (ArishSultan) [reply ubuntu_desktop_installer/lib/welcomepage.dart:41]: I figured it out, `setupAppLocalizations` was not called in test initializations.

So, the `Intl.defaultLocale` gave a null value which lead to the `locale` variable being null too.

**Comment by oSoMoN** [comment]
The "translations" section of README.md probably needs to be updated to reflect the new workflow.

**Comment by oSoMoN** [comment]
I'm seeing a couple of functional regressions:
 - when I start the app with French as the default locale, French is correctly selected by default in the language list, but the French translations aren't loaded (the UI is in English)
 - the continue button on the "try or install" screen should be disabled by default, until one of the three options is selected

**Comment by oSoMoN** [comment]
Note: I just pushed a couple of commits that address the two functional regressions I was mentioning in my previous comment. I'm continuing with the review.

File: ubuntu_desktop_installer/lib/l10n/app_localizations.dart [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        shortMonthDayFormat: DateFormat('MMM d'),
+      ),
+    );
+  }
+
+  @override
+  bool shouldReload(_UbuntuLocalizationsDelegate old) => false;
+}
+
+class GlobalUbuntuLocalizations extends MaterialLocalizationEn {
```
Comment (oSoMoN) [inline ubuntu_desktop_installer/lib/l10n/app_localizations.dart:300]: This approach doesn't look right: some code appears to be copy/pasted from an implementation of the 'nn' locale (Norwegian Nynorsk), and extending `MaterialLocalizationEn` [isn't recommended](https://api.flutter.dev/flutter/flutter_localizations/GlobalMaterialLocalizations-class.html#overriding-translations).

For occitan support, maybe we can extend [`MaterialLocalizationCa`](https://api.flutter.dev/flutter/flutter_localizations/MaterialLocalizationCa-class.html)? @Mejans what do you think? Would that be close enough with regards to date/time and number formats?

**Review by oSoMoN — CHANGES_REQUESTED** [review]

File: ubuntu_desktop_installer/lib/l10n/app_localizations.dart (hunk 2 above)
Comment (Mejans) [reply ubuntu_desktop_installer/lib/l10n/app_localizations.dart:300]: yes, linguistically one language, politically two :) So yes it would be fine.

Comment (ArishSultan) [reply ubuntu_desktop_installer/lib/l10n/app_localizations.dart:300]: No problem, As I said it was a dummy implementation (a quick fix).

**Comment by MarcusTomlinson** [comment]
So sorry for the CI noise. The CLA check was going bonkers. Looks to be fixed now.

**Comment by oSoMoN** [comment]
> The "translations" section of README.md probably needs to be updated to reflect the new workflow.

While doing that, I've been wondering how to automate the extraction of strings from source code to the translation template. I haven't seen anything related to that mentioned in the documentation. Not a blocker, so I'm going to go ahead and merge this PR now, but it would be a nice improvement for a future PR.

**Pull request merged**